src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs

```diff
@@ -701,36 +701,37 @@ public ArrayTypeSymbol Retarget(ArrayTypeSymbol type)
 
             internal ImmutableArray<CustomModifier> RetargetModifiers(ImmutableArray<CustomModifier> oldModifiers, out bool modifiersHaveChanged)
             {
-                int i;
-                int count = oldModifiers.Length;
                 modifiersHaveChanged = false;
+                ArrayBuilder<CustomModifier> newModifiers = null;
 
-                if (count != 0)
+                for (int i = 0; i < oldModifiers.Length; i++)
                 {
-                    CustomModifier[] newModifiers = new CustomModifier[count];
+                    var oldModifier = oldModifiers[i];
+                    NamedTypeSymbol newModifier = Retarget((NamedTypeSymbol)oldModifier.Modifier, RetargetOptions.RetargetPrimitiveTypesByName); // should be retargeted by name
 
-                    for (i = 0; i < count; i++)
+                    if (!newModifier.Equals(oldModifier.Modifier))
                     {
-                        var oldModifier = oldModifiers[i];
-                        NamedTypeSymbol newModifier = Retarget((NamedTypeSymbol)oldModifier.Modifier, RetargetOptions.RetargetPrimitiveTypesByName); // should be retargeted by name
+                        modifiersHaveChanged = true;
 
-                        if (!newModifier.Equals(oldModifier.Modifier))
+                        if (newModifiers == null)
                         {
-                            modifiersHaveChanged = true;
-                            newModifiers[i] = oldModifier.IsOptional ?
-                                                CSharpCustomModifier.CreateOptional(newModifier) :
-                                                CSharpCustomModifier.CreateRequired(newModifier);
+                            newModifiers = ArrayBuilder<CustomModifier>.GetInstance(oldModifiers.Length);
+                            newModifiers.AddRange(oldModifiers, i);
                         }
-                        else
-                        {
-                            newModifiers[i] = oldModifier;
-                        }
-                    }
 
-                    return newModifiers.AsImmutableOrNull();
+                        newModifiers.Add(oldModifier.IsOptional ?
+                                            CSharpCustomModifier.CreateOptional(newModifier) :
+                                            CSharpCustomModifier.CreateRequired(newModifier));
+                    }
+                    else if (newModifiers != null) 
+                    {
+                        newModifiers.Add(oldModifier);
+                    }
                 }
 
-                return oldModifiers;
+                Debug.Assert(newModifiers == null || newModifiers.Count == oldModifiers.Length);
+                Debug.Assert((newModifiers != null) == modifiersHaveChanged);
```

Comment:
Minor point: We could assign `modifiersHaveChanged = (newModifiers != null);` here instead of the two assignments above. Then the assert is unnecessary.